File: cohere/typescript/src/services/chat/validation.ts
/**
 * Validation for chat requests.
 */

import { ValidationError } from '../../errors';
import type { ValidationDetail } from '../../errors';
import type { ChatRequest, ChatMessage, Tool } from './types';

/**
 * Validate a chat request
 */
export function validateChatRequest(request: ChatRequest): void {
  const errors: ValidationDetail[] = [];

  // Validate message
  if (!request.message || request.message.trim() === '') {
    errors.push({
      field: 'message',
      message: 'Message is required and cannot be empty',
      code: 'REQUIRED',
    });
  }

  // Validate temperature
  if (request.temperature !== undefined) {
    if (request.temperature < 0 || request.temperature > 2) {
      errors.push({
        field: 'temperature',
        message: 'Temperature must be between 0 and 2',
        code: 'OUT_OF_RANGE',
      });
    }
  }

  // Validate top_p
  if (request.topP !== undefined) {
    if (request.topP < 0 || request.topP > 1) {
      errors.push({
        field: 'topP',
        message: 'top_p must be between 0 and 1',
        code: 'OUT_OF_RANGE',
      });
    }
  }

  // Validate top_k
  if (request.topK !== undefined) {
    if (request.topK < 0 || request.topK > 500) {
      errors.push({
        field: 'topK',
        message: 'top_k must be between 0 and 500',
        code: 'OUT_OF_RANGE',
      });
    }
  }

  // Validate max_tokens
  if (request.maxTokens !== undefined) {
    if (request.maxTokens < 1 || request.maxTokens > 4096) {
      errors.push({
        field: 'maxTokens',
        message: 'max_tokens must be between 1 and 4096',
        code: 'OUT_OF_RANGE',
      });
    }
  }

  // Validate frequency_penalty
  if (request.frequencyPenalty !== undefined) {
    if (request.frequencyPenalty < 0 || request.frequencyPenalty > 1) {
      errors.push({
        field: 'frequencyPenalty',
        message: 'frequency_penalty must be between 0 and 1',
        code: 'OUT_OF_RANGE',
      });
    }
  }

  // Validate presence_penalty
  if (request.presencePenalty !== undefined) {
    if (request.presencePenalty < 0 || request.presencePenalty > 1) {
      errors.push({
        field: 'presencePenalty',
        message: 'presence_penalty must be between 0 and 1',
        code: 'OUT_OF_RANGE',
      });
    }
  }

  // Validate chat history
  if (request.chatHistory) {
    const historyErrors = validateChatHistory(request.chatHistory);
    errors.push(...historyErrors);
  }

  // Validate tools
  if (request.tools) {
    const toolErrors = validateTools(request.tools);
    errors.push(...toolErrors);
  }

  // Validate tool results require tools
  if (request.toolResults && request.toolResults.length > 0 && !request.tools) {
    errors.push({
      field: 'toolResults',
      message: 'Tool results require tools to be defined',
      code: 'INVALID_COMBINATION',
    });
  }

  // Validate documents
  if (request.documents) {
    for (let i = 0; i < request.documents.length; i++) {
      const doc = request.documents[i];
      if (!doc?.text || doc.text.trim() === '') {
        errors.push({
          field: `documents[${i}].text`,
          message: 'Document text is required',
          code: 'REQUIRED',
        });
      }
    }
  }

  // Throw if there are validation errors
  if (errors.length > 0) {
    throw new ValidationError('Invalid chat request', errors);
  }
}

/**
 * Validate chat history
 */
function validateChatHistory(history: ChatMessage[]): ValidationDetail[] {
  const errors: ValidationDetail[] = [];

  for (let i = 0; i < history.length; i++) {
    const message = history[i];

    if (!message) {
      errors.push({
        field: `chatHistory[${i}]`,
        message: 'Message cannot be null',
        code: 'REQUIRED',
      });
      continue;
    }

    if (!message.role) {
      errors.push({
        field: `chatHistory[${i}].role`,
        message: 'Message role is required',
        code: 'REQUIRED',
      });
    }

    if (!message.content || message.content.trim() === '') {
      errors.push({
        field: `chatHistory[${i}].content`,
        message: 'Message content is required',
        code: 'REQUIRED',
      });
    }

    // Validate tool messages have tool_call_id
    if (message.role === 'TOOL' && !message.toolCallId) {
      errors.push({
        field: `chatHistory[${i}].toolCallId`,
        message: 'Tool messages require toolCallId',
        code: 'REQUIRED',
      });
    }
  }

  return errors;
}

/**
 * Validate tools
 */
function validateTools(tools: Tool[]): ValidationDetail[] {
  const errors: ValidationDetail[] = [];
  const toolNames = new Set<string>();

  for (let i = 0; i < tools.length; i++) {
    const tool = tools[i];

    if (!tool) {
      errors.push({
        field: `tools[${i}]`,
        message: 'Tool cannot be null',
        code: 'REQUIRED',
      });
      continue;
    }

    if (!tool.name || tool.name.trim() === '') {
      errors.push({
        field: `tools[${i}].name`,
        message: 'Tool name is required',
        code: 'REQUIRED',
      });
    } else if (toolNames.has(tool.name)) {
      errors.push({
        field: `tools[${i}].name`,
        message: `Duplicate tool name: ${tool.name}`,
        code: 'DUPLICATE',
      });
    } else {
      toolNames.add(tool.name);
    }

    if (!tool.description || tool.description.trim() === '') {
      errors.push({
        field: `tools[${i}].description`,
        message: 'Tool description is required',
        code: 'REQUIRED',
      });
    }

    // Validate parameters
    if (tool.parameters) {
      const paramNames = new Set<string>();
      for (let j = 0; j < tool.parameters.length; j++) {
        const param = tool.parameters[j];

        if (!param) continue;

        if (!param.name) {
          errors.push({
            field: `tools[${i}].parameters[${j}].name`,
            message: 'Parameter name is required',
            code: 'REQUIRED',
          });
        } else if (paramNames.has(param.name)) {
          errors.push({
            field: `tools[${i}].parameters[${j}].name`,
            message: `Duplicate parameter name: ${param.name}`,
            code: 'DUPLICATE',
          });
        } else {
          paramNames.add(param.name);
        }
      }
    }
  }

  return errors;
}

File: cohere/rust/src/services/chat/validation.rs
//! Validation for chat requests.

use super::types::ChatRequest;
use crate::errors::{CohereError, CohereResult, ValidationDetail};

/// Validate a chat request
pub fn validate_chat_request(request: &ChatRequest) -> CohereResult<()> {
    let mut errors = Vec::new();

    // Validate message
    if request.message.is_empty() {
        errors.push(ValidationDetail::new("message", "Message cannot be empty"));
    }

    // Validate temperature
    if let Some(temp) = request.temperature {
        if !(0.0..=5.0).contains(&temp) {
            errors.push(ValidationDetail::with_value(
                "temperature",
                "Temperature must be between 0.0 and 5.0",
                temp.to_string(),
            ));
        }
    }

    // Validate max_tokens
    if let Some(max) = request.max_tokens {
        if max == 0 {
            errors.push(ValidationDetail::with_value(
                "max_tokens",
                "max_tokens must be greater than 0",
                max.to_string(),
            ));
        }
        if max > 4096 {
            errors.push(ValidationDetail::with_value(
                "max_tokens",
                "max_tokens cannot exceed 4096",
                max.to_string(),
            ));
        }
    }

    // Validate top-k
    if let Some(k) = request.k {
        if k == 0 {
            errors.push(ValidationDetail::with_value(
                "k",
                "k must be greater than 0",
                k.to_string(),
            ));
        }
    }

    // Validate top-p
    if let Some(p) = request.p {
        if !(0.0..=1.0).contains(&p) {
            errors.push(ValidationDetail::with_value(
                "p",
                "p must be between 0.0 and 1.0",
                p.to_string(),
            ));
        }
    }

    // Validate frequency_penalty
    if let Some(fp) = request.frequency_penalty {
        if !(-2.0..=2.0).contains(&fp) {
            errors.push(ValidationDetail::with_value(
                "frequency_penalty",
                "frequency_penalty must be between -2.0 and 2.0",
                fp.to_string(),
            ));
        }
    }

    // Validate presence_penalty
    if let Some(pp) = request.presence_penalty {
        if !(-2.0..=2.0).contains(&pp) {
            errors.push(ValidationDetail::with_value(
                "presence_penalty",
                "presence_penalty must be between -2.0 and 2.0",
                pp.to_string(),
            ));
        }
    }

    // Validate stop sequences
    if let Some(ref seqs) = request.stop_sequences {
        if seqs.len() > 5 {
            errors.push(ValidationDetail::with_value(
                "stop_sequences",
                "Cannot have more than 5 stop sequences",
                seqs.len().to_string(),
            ));
        }
    }

    // Validate tools
    if let Some(ref tools) = request.tools {
        for (i, tool) in tools.iter().enumerate() {
            if tool.name.is_empty() {
                errors.push(ValidationDetail::new(
                    format!("tools[{}].name", i),
                    "Tool name cannot be empty",
                ));
            }
            if tool.description.is_empty() {
                errors.push(ValidationDetail::new(
                    format!("tools[{}].description", i),
                    "Tool description cannot be empty",
                ));
            }
        }
    }

    if errors.is_empty() {
        Ok(())
    } else {
        Err(CohereError::Validation {
            message: format!("Chat request validation failed: {} error(s)", errors.len()),
            details: errors,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_validate_valid_request() {
        let request = ChatRequest::new("Hello, how are you?");
        assert!(validate_chat_request(&request).is_ok());
    }

    #[test]
    fn test_validate_empty_message() {
        let request = ChatRequest::new("");
        let result = validate_chat_request(&request);
        assert!(result.is_err());
    }

    #[test]
    fn test_validate_invalid_temperature() {
        let mut request = ChatRequest::new("Hello");
        request.temperature = Some(10.0);
        let result = validate_chat_request(&request);
        assert!(result.is_err());
    }

    #[test]
    fn test_validate_invalid_max_tokens() {
        let mut request = ChatRequest::new("Hello");
        request.max_tokens = Some(0);
        let result = validate_chat_request(&request);
        assert!(result.is_err());
    }

    #[test]
    fn test_validate_invalid_p() {
        let mut request = ChatRequest::new("Hello");
        request.p = Some(1.5);
        let result = validate_chat_request(&request);
        assert!(result.is_err());
    }
}
